Core/include/Acts/Clusterization/HkClusterization.ipp

```diff
@@ -0,0 +1,233 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2022 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include <array>
+#include <vector>
+
+#include <boost/pending/disjoint_sets.hpp>
+
+namespace Acts {
+namespace internal {
+
+// Machinery for validating generic Cell/Cluster types at compile-time
+
+template <typename...>
+using void_t = void;
+
+template <typename, typename T = void>
+struct cell_type_has_required_functions : std::false_type {};
+
+template <typename T>
+struct cell_type_has_required_functions<
+    T, void_t<decltype(get_cell_row(std::declval<T>())),
+              decltype(get_cell_column(std::declval<T>())),
+              decltype(get_cell_label(std::declval<T&>()))>>
+    : std::true_type {};
+
+template <typename, typename, typename T = void>
+struct cluster_type_has_required_functions : std::false_type {};
+
+template <typename T, typename U>
+struct cluster_type_has_required_functions<
+    T, U,
+    void_t<decltype(cluster_add_cell(std::declval<T>(), std::declval<U>()))>>
+    : std::true_type {};
+
+template <typename T>
+constexpr bool CellConcept = cell_type_has_required_functions<T>();
+
+template <typename T, typename U>
+constexpr bool ClusterConcept = cluster_type_has_required_functions<T, U>();
+
+#define CHECK_CELL_TYPE(T)                                                  \
+  do {                                                                      \
+    static_assert(Acts::internal::CellConcept<T>,                           \
+                  "Cell type should have the following functions:'int "     \
+                  "get_cell_row(const Cell&)', 'int get_cell_column(const " \
+                  "Cell&)', 'Label& get_cell_label(Cell&)'");         \
+  } while (0)
+
+#define CHECK_CLUSTER_TYPE(T, U)                                           \
+  do {                                                                     \
+    static_assert(Acts::internal::ClusterConcept<T, U>,                    \
+                  "Cluster type should have the following function:'void " \
+                  "cluster_add_cell(Cluster&, const Cell&)'");             \
+  } while (0)
```

Comment:
If you don't use `__LINE__` or `__FILE__`, could these just be `constexpr` functions?

Reply:
`constexpr` are not garanteed to actually be evaluated at compile time right? I'd prefer keeping these as macros since we need a compile time check (not sure how that interacts with static_assert...)

However, these macros should be undef'd at the end of the file so they don't leak into the calling code, I'll add that.

Reply:
I'm pretty sure if you plug these `static_assert`s into a template function, they work identically to the macros here.

Reply:
Okay, I will investigate -- I should learn more about the compile time execution mechanisms anyways...

Reply:
Done! It's indeed much cleaner now.